Punishments: Add battle ban

AustinXII opened the pull request:
I was bored and this has been requested forever.

AustinXII commented:
Oh, my user.js commits didn't get pushed, hold up.

AustinXII commented:
someone will probably have to explain the typescript errors though,,

AustinXII commented:
also travisbuddy pls stop spamming my mistyping, I get it D:

HoeenCoder commented on chat-commands.js:
```diff
@@ -2479,6 +2479,59 @@ exports.commands = {
+		if (!targetUser) {
+			return this.errorReply(`User '${this.targetUsername}' not found.`);
+		}
+		if (!this.can('forcerename', targetUser)) return false;
```
Why not `lock` assuming this is going to be for %+ and not @+ (iirc the policy disscusion never finished)

AustinXII replied:
I copied most of this code from a different command, yeah I can change it to lock

AustinXII replied:
Forcerename is available to drivers so this is %+. Yeah the policy wasn’t finalized, I just did it like this since it’s the one I most agreed with. Gdrivers deserve battle moderation as well as server moderation imo. It seems strange to restrict global moderation from global staff.

panpawn replied:
Or, how about we don't overlap permissions and just create new ones for new features, that works too

AustinXII replied:
Why? is this preferred now a days?

HoeenCoder commented on chat-commands.js:
```diff
@@ -2479,6 +2479,59 @@ exports.commands = {
+		const battlebanMessage = `${targetUser.name} was banned from starting new games by ${user.name}${reasonText}`;
+		this.privateModAction(`(${battlebanMessage})`);
+
+		// Notify staff room when a user is locked outside of it.
```
not lock

HoeenCoder commented on chat-commands.js:
```diff
@@ -2479,6 +2479,59 @@ exports.commands = {
+			Rooms('staff').addByUser(user, "<<" + room.id + ">> " + battlebanMessage);
+		}
+
+		this.globalModlog("BATTLEBANNED", targetUser, ` by ${user.userid}${reasonText}`);
```
Nitpicky but i think `BATTLE BAN` would be better

AustinXII replied:
It’s NAMELOCKED so I disagree

AustinXII replied:
Sorry it’s NAMELOCK, I could change it to BATTLEBAN without the space tho

HoeenCoder commented on chat-commands.js:
```diff
@@ -2479,6 +2479,59 @@ exports.commands = {
+
+		if (letplay) {
+			this.addModAction(letplay + " was allowed to battle again by " + user.name + "." + reason);
+			if (!reason) this.globalModlog("UNBATTLEBAN", target, " by " + user.userid);
```
Continuing with nitpickness `UNBATTLE BAN`

HoeenCoder commented on punishments.js:
```diff
@@ -843,6 +845,56 @@ Punishments.unnamelock = function (name) {
+ * @param {...string} [reason]
+ * @return {PunishmentRow}
+ */
+Punishments.battleban = function (user, expireTime, id, ...reason) {
```
I'm pretty sure a user can only have 1 global punishment at a time.

I feel that having a locked user also being battle banned should be allowed, though it might be tricky in this case. We may have to just global ban in those situations.

AustinXII replied:
oh I didn't know that. Yeah I wanted it so battle ban and lock were separate so users could still chat in chatrooms if they got battlebanned, but I suppose you should just be locked and battlebanned instead if you're bad enough to warrant one.

AustinXII replied:
Although having it done the intended way would be preferred, how would I go about doing that ?

panpawn replied:
Battle bans should act like namelocks where they also cannot talk, that could solve this issue

AustinXII replied:
That’s what we said, is that how everyone feels? Hoeen and I both agree there’s a reason to be banned from battling and not locked at the same time. But honestly I’d be fine with either option

Zarel commented:
This overlaps with lock too much to be a global punishment... Is it really necessary?

AustinXII commented:
@Zarel I don’t know if it’s necessary but even you have to admit that you have seen this request a lot and there must be reason for it. With all the complaints staff have had over not being able to do anything about timerstallers that have gone unresolved, I feel this is the best solution. I’m sure there could be merit for this command to be leader+ as well although at a much lesser scale.

panpawn commented:
>mfw comments again if you squash

If you tested ``npm test`` before squashing and pushing a new commit, then it wouldn't comment again when the new commit fails

Zarel commented:
I don't _want_ staff to do anything against timerstallers. It's too fraught with complication. A battle ban is a global ban. Just call a moderator to `/gban` if you _really_ want to. But for timerstallers, the correct thing to do is always "nothing", from an individual enforcement level.

The only real solution you can do is change game rules. Things like "You can't start a new game while timerstalling", etc.

AustinXII commented:
@panpawn I don’t know how to fix the last error as stated previously, I squashed my commits that fixed the other errors.

panpawn commented on punishments.js:
```diff
@@ -24,14 +24,15 @@ const SHAREDIPS_FILE = 'config/sharedips.tsv';
 const PUNISH_TRUSTED = false;
 
-const PUNISHMENT_POINT_VALUES = {MUTE: 2, BLACKLIST: 3, ROOMBAN: 4};
+const PUNISHMENT_POINT_VALUES = {MUTE: 2, BLACKLIST: 3, BATTLEBAN: 4, ROOMBAN: 4};
```
???

I'm confused, is this a global punishment or a room punishment? I don't think it should count towards the monitor like this if its a global punishment.

AustinXII replied:
It’s a global punishment but it’s counting towards a lock if they cause other issues

AustinXII replied:
But yeah if we don’t change it so battle bans auto lock, this is undoubtably needed.

Zarel replied:
In its present form, I agree that this should be counted.

panpawn commented on punishments.js:
```diff
@@ -171,6 +172,7 @@ Punishments.punishmentTypes = new Map([
 	['LOCK', 'locked'],
 	['BAN', 'globally banned'],
 	['NAMELOCK', 'namelocked'],
+	['BATTLEBAN', 'battleBanned'],
```
Why camel case? see global ban casing here

AustinXII replied:
Personal habit I figured someone was going to point it but only realized after finishing.

panpawn commented on punishments.js:
```diff
@@ -843,6 +845,56 @@ Punishments.unnamelock = function (name) {
+	if (!expireTime) expireTime = Date.now() + BATTLEBAN_DURATION;
+	let punishment = ['BATTLEBANNED', id, expireTime, ...reason];
+
+	let affected = Punishments.punish(user, punishment);
```
``const``

panpawn commented on punishments.js:
```diff
@@ -843,6 +845,56 @@ Punishments.unnamelock = function (name) {
+ * @param {string} name
+ */
+Punishments.unbattleban = function (name) {
+	let user = Users(name);
```
``const``

panpawn commented on punishments.js:
```diff
@@ -843,6 +845,56 @@ Punishments.unnamelock = function (name) {
+ */
+Punishments.unbattleban = function (name) {
+	let user = Users(name);
+	let id = toId(name);
```
``const``

AustinXII replied:
id gets manipulated

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,59 @@ exports.commands = {
+			return this.errorReply(`User '${this.targetUsername}' not found.`);
+		}
+		if (!this.can('forcerename', targetUser)) return false;
+		if (targetUser.battleBanned) return this.errorReply(`User '${targetUser.name}' is already banned from battles.`);
```
... What does this even mean? Their roombanned from battles? This would be worded better

AustinXII replied:
“From battling” is what I meant

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,59 @@ exports.commands = {
+			reason = ' (' + targetUser.battleBanned + ')';
+		}
+
+		let letplay = Punishments.unbattleban(target);
```
This is a pretty bad variable name, what about ``const unbanned`` or ``const canBattle``?

panpawn replied:
Speaking of which, we should probably introduce a ``this.canBattle`` function if this feature does happen

panpawn commented on punishments.js:
```diff
@@ -843,6 +845,56 @@ Punishments.unnamelock = function (name) {
+	let id = toId(name);
+	/** @type {string[]} */
+	let success = [];
+	let unpunished = Punishments.unpunish(name, 'BATTLEBAN');
```
You use ``BATTLEBAN`` in some places and ``BATTLEBANNED`` in others - pick one to consistently use

AustinXII commented:
@Zarel yeah I was just making a proposal, I knew there was a good chance it wouldn’t be approved. I think a battle ban instead of globally banning would have use though. I don’t think many staff would agree a ban from the entire server is necessary for timer stalling but do think many would agree banning them from new battles would be warranted. But yeah If you prefer we stick to how we currently deal with timesralling I'm fine with this being closed.

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,59 @@ exports.commands = {
+	battleban: function (target, room, user) {
+		if (!target) return this.parse('/help battleban');
+
+		let reason = this.splitTarget(target, true);
```
I think this should require a reason like global ban.

AustinXII commented on chat-commands.js:
```diff
@@ -3623,6 +3676,9 @@ exports.commands = {
 			return this.popupReply(`You are locked and cannot challenge unlocked users.`);
 		}
+		if (user.battleBanned) {
+			return this.popupReply(`You are battle banned and cannot challenge users.`);
```
True, I just copied how lock was worded

AustinXII commented:
@panpawn where were you for the punishment refactor 90% of these errors are from me copying the code style from the rest of the code in the files D: lol but yeah I’ll fix it if @Zarel changes his mind about wanting this.

Zarel commented:
Honestly, this half-TypeScript is even harder to maintain than full TypeScript.

But making it so servers don't have to worry about TypeScript is possibly an advantage?

Zarel commented:
Anyway, I would probably implement this as a room punishment in roomid `'battle'`.

panpawn commented:
>Anyway, I would probably implement this as a room punishment in roomid 'battle'.

What about the global room?... No?

AustinXII commented:
I still don't know how to fix that TS error but I think everything else is fine

panpawn commented on chat-commands.js:
```diff
@@ -3676,8 +3679,8 @@ exports.commands = {
-		if (user.battleBanned) {
-			return this.popupReply(`You are battle banned and cannot challenge users.`);
+		if (user.battlebanned) {
+			return this.popupReply(`You are banned from battling and cannot challenge users.`);
```
I still think this should say the reason, if there was one

panpawn replied:
Nevermind, reasons are required - this should say it then

AustinXII replied:
I guess I should just do that for all of them

AustinXII commented:
Also hoeen's concerns were inaccurate, these punishments can stack and imo this is the best way to implement it. I don't overly get what you mean @Zarel

panpawn commented:
He means that this should be handled in an abstract room like the Global room, or a new one with the id ``battle`` instead of handling it as a global punishment

AustinXII commented:
ah I assumed that's what he meant, although I don't know if I know PS well enough to do that easily. Is there any advantage to doing it that way over this way? I think having battles be listed on a global scale is appropriate as it takes away your ability to battle anywhere on the server, although I almost forgot to handle this in ``tournaments.js``

Zarel commented:
The main reason is because I thought global punishments couldn't stack. Was that changed at some point?

Zarel commented on users.js:
```diff
@@ -465,6 +465,7 @@ class User {
 		/** @type {?false | string} */
 		this.semilocked = false;
 		this.namelocked = false;
+		this.battlebanned = false;
```
Your problem is here. What is this variable supposed to be?

I would probably make it a pure string. `this.battlebanned = '';`

Zarel replied:
(Specifically, the reason TypeScript is yelling at you is here.)

Zarel replied:
So, first: Whenever you define a variable, you should keep in mind its contract: what it means. Here is your intended contract for this variable:

- `false`: not battle banned
- lowercase alphanumeric string: battle banned under that userid
- anything else: bug

In TypeScript, you would declare that this way:

```
/**
 * `false` if not battle banned, otherwise battle-banned userid
 * @type {string | false}
 */
this.battlebanned = false;
```

Zarel replied:
I'm instead proposing this contract:

- `''`: not battle banned
- lowercase alphanumeric string: battle banned under that userid
- anything else: bug

You would declare that this way:

```
/** '' if not battle banned, otherwise battle-banned userid */
this.battlebanned = '';
```

Because you're only using one type now, it can be inferred, so you don't need `@type`.

Zarel replied:
Note that your original contract is against PS's modern coding standards, anyway, see here:

https://github.com/Zarel/Pokemon-Showdown/blob/master/CONTRIBUTING.md#optionals-null-vs-undefined-vs-false

AustinXII replied:
Ah I see, I should have looked at how lock instead of namelock, I would have seen the typescript stuff lol. Also I’ll update the code standard stuff, I should have made sure the current code in the file that I was using to guide my code styling was following the contributing standards before starting lol

Zarel commented:
For the record, it's not your fault that your code wasn't up to modern code standards.

I'm 100% okay with you copying the same standard as the code around you; I do _not_ expect you to have CONTRIBUTING memorized (although I would appreciate it! especially since you lead damagecalc).

AustinXII commented:
Yeah I’ll have to look over it again, there’s definitely been changes since the last time I looked. I don’t pay __too__ much attention to those types of changes for ps but I do try stay up to date with most coding changes. Is there someone working on changing the code to make it on par with the standards or do we just update as we go? I could probably work on converting some files soon so there are less mixups like this in the future

Zarel commented:
Mostly, we just update as we go. I put it in "Ideas for New Developers" and sometimes I ask people to do it, but it's not consistent.

AustinXII commented:
sorry for the delay @Zarel

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,62 @@ exports.commands = {
 	},
 	blacklisthelp: [`/blacklist [username], [reason] - Blacklists the user from the room you are in for a year. Requires: # & ~`],
 
+	bb: 'battleban',
```
I don't know how I feel about this alias. Does this really need an alias?

AustinXII replied:
Doesn’t matter to me but why not?

panpawn replied:
Because it's an unclear alias. I think this command isn't going to be used often enough to warrant it, and its not hard to type out ``/battleban``. Think of how ``/b`` was removed to global ban, because people thought it might do something else. Same idea here.

AustinXII replied:
Yeah I already removed it, although I don’t think it’s “unclear.” It’s definitely not going to be used enough to warrant an alias tho. Also “b” was removed from globalban because of the serverity of the punishment and because it was changed over to an alias for /roomban. If this had more use I would definitely argue /bb is fine.

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,62 @@ exports.commands = {
+		if (!target) {
+			return this.errorReply("Battle banning require a reason.");
+		}
+		if (!this.can('lock', targetUser)) return false;
```
``if (!this.can('lock', targetUser)) return;``

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,62 @@ exports.commands = {
+
+		// Notify staff room when a user is banned from battling outside of it.
+		if (room.id !== 'staff' && Rooms('staff')) {
+			Rooms('staff').addByUser(user, "<<" + room.id + ">> " + battlebanMessage);
```
template string

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,62 @@ exports.commands = {
+		this.globalModlog("BATTLEBAN", targetUser, ` by ${user.userid}${reasonText}`);
+		Ladders.cancelSearches(targetUser);
+		Punishments.battleban(targetUser, null, null, target);
+		targetUser.popup(`|modal|${user.name} has prevented you from starting new battles${reasonText}`);
```
I still think this should say how long its for

AustinXII replied:
Oh sorry I forgot about that

panpawn commented:
Does this modlog the user's IP? I think it should modlog the user's IP.

AustinXII commented:
@panpawn yes this.globalModlog logs IPs

HoeenCoder commented:
I talked to @bumbadadabum to confirm that a user can only have 1 global punishment at a time. Thats why applying this to the global room or a room with the id 'battle' would be nessecary.

AustinXII commented:
Whoops. Also what? @HoeenCoder Feel free to test this code it worked for me?

Hidden50 commented on chat-commands.js:
```diff
@@ -2479,6 +2479,60 @@ exports.commands = {
+			return this.errorReply(`The reason is too long. It cannot exceed ${MAX_REASON_LENGTH} characters.`);
+		}
+		if (!target) {
+			return this.errorReply("Battle banning require a reason.");
```
grammar; this should be "Battle bans require a reason."

HoeenCoder commented:
It only partially works with multiple global punishments actually. For example, if a user is battlebanned while locked it is logged as a lock not a battle ban in punishments.tsv, (meaning the next time thier user object expires they arent battle banned anymore). Also unlocking a user removes thier battle ban from punishments.tsv as well.

Hidden50 commented on chat-commands.js:
```diff
@@ -2479,6 +2479,60 @@ exports.commands = {
+		let targetUser = Users.get(target);
+		let reason = '';
+		if (targetUser && targetUser.battlebanned) {
+			reason = ' (`${targetUser.battlebanned}`)';
```
there's a template string inside a regular string; should be  `` `(${targetUser.battlebanned})` ``

panpawn commented:
One problem about doing this and making it a room punishment is that they won't be persistent thru server restarts. Is this an issue we should worry about?

AustinXII commented:
@HoeenCoder @panpawn yeah that is an issue, does implementing this the same way as namelocks fix this? (locking the user aswell.) I don’t think an inconsistent battle ban is good enough tbh. Although Hoeen did you test that or assuming that? I swear after killing the server while I was locked and battled banned it booted up the punishments as intended but I could be wrong / screwed something up with eval . No clue of the unlocking thing though, I didn’t test this to that extent

Hidden50 commented:
> does implementing this the same way as namelocks fix this?

The current global punishments are strictly hierarchical: lock -> namelock -> gban. Your user object can have multiple of them, but in the punishments file they replace each other in that order. The next time your user object expires and you log back in, you will only have one.

This is also the case with battleban, except I think we want it to be combinable with namelocks. Battlebanning should not remove a namelock, and namelocking should not remove a battleban.

AustinXII commented:
@hidden50 Uh you mean regular locks right? although I guess they’re pretty much the same thing. That makes sense about the hierarchy and what I assumed, being overwritten by a worse punishment is intended to an extent although sadly it’s passing the line. But yeah I agree, I think most staff can agree being combinable with lock + namelock is probably best so I’ll have to work on that

Zarel commented:
namelock + battleban is literally just a globalban. Just gban them.

Hidden50 commented:
> namelock + battleban is literally just a globalban. Just gban them.

I considered that, but should it be automatic? Should punishments.namelock apply gban to a user who is already battlebanned, and the other way around?

If not, what should happen instead? If the code doesn't cover it, this would silently remove the other punishment.

There is also the following difference: A user that is namelocked and battlebanned can appeal their punishment on PS. If they are globally banned, they have to go through smogon.

Zarel commented:
We only support one global punishment at a time. Battle ban should be a room punishment, so it doesn't override existing global punishments. I already explained this. :/

HoeenCoder commented:
@AustinXII I tested the interactions between locks (not namelocks) and battle bans with Orda-Y, and I looked at punishments.tsv to see what was being saved. It definetly was not working properly.

We should either make punishments support multiple global punishments or (the easier way) make battle bans a room punishment on global or a room called `"battle"`. 

Also @panpawn all room punishments but mutes will not expire when the server restarts (they are saved). https://github.com/Zarel/Pokemon-Showdown/blob/master/punishments.js#L292

Zarel commented:
For the record, I already said that it _must_ be on the roomid `"battle"`. If you want to do it any other way, you'll need to talk it through with me first.

AustinXII commented:
Whoops I did not mean to push that yet

HoeenCoder commented:
Talked with @AustinXII about this and offered to help out (hence the commit).
This isn't completely tested yet, I've pushed for code review purposes.

AustinXII commented:
@HoeenCoder just curious what haven't you tested? Everything seems to be working as expected when testing.

HoeenCoder commented:
Interaction between battle bans and other global punishments (mostly locks)

AustinXII commented:
ah yeah that was the first thing I test lol, it seems to work fine

panpawn commented on tournaments/index.js:
```diff
@@ -281,7 +281,7 @@ class Tournament {
 
 		if (user.battlebanned) {
-			output.sendReply('|tournament|error|Battle Banned');
+			output.sendReply('|tournament|error|BattleBanned');
```
Is this supported in the client?

AustinXII commented:
@Zarel this appears to be ready for your final review.

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,58 @@ exports.commands = {
+		if (!target) return this.parse(`/help battleban`);
+
+		target = this.splitTarget(target);
+		let targetUser = this.targetUser;
```
``const``

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,58 @@ exports.commands = {
+	battleban: function (target, room, user, connection) {
+		if (!target) return this.parse(`/help battleban`);
+
+		target = this.splitTarget(target);
```
It'd be more readable if this was a separate variable; I suggest ``reason``

AustinXII replied:
I suppose that are other commands that define it as target although :P

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,58 @@ exports.commands = {
+		targetUser.popup(`|modal|${user.name} has prevented you from starting new battles for 2 days${reasonText}`);
+		return true;
+	},
+	battlebanhelp: [`/battleban [username], [reason] - Prevents the user from starting new battles and shows them the [reason]. Requires: % @ * & ~`],
```
The help command should say that its for 2 days

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,58 @@ exports.commands = {
+		if (!target) return this.parse('/help unbattleban');
+		if (!this.can('lock')) return;
+
+		let targetUser = Users.get(target);
```
``const``

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,58 @@ exports.commands = {
+		if (!this.can('lock')) return;
+
+		let targetUser = Users.get(target);
+		let unbanned = Punishments.unbattleban(target);
```
``const``

panpawn replied:
(also the line above, too)

panpawn commented on chat-commands.js:
```diff
@@ -2479,6 +2479,58 @@ exports.commands = {
+
+		this.globalModlog("BATTLEBAN", targetUser, ` by ${user.userid}${reasonText}`);
+		Ladders.cancelSearches(targetUser);
+		Punishments.battleban(targetUser, null, null, target);
```
``target`` should be ``reason`` here

AustinXII replied:
whoops

AustinXII replied:
missed that one

panpawn commented:
I'm confused how thoroughly this has been tested or not, can someone clarify?

AustinXII commented:
@panpawn I’ve tested everything, that’s why I said “final review”, I only said appears because I thought there were no more minor corrections needed :P

Zarel commented on ladders.js:
```diff
@@ -97,6 +97,10 @@ class Ladder extends LadderStore {
 			connection.popup(message);
 			return null;
 		}
+		if (user.battlebanned) {
```
There's no need to have a `.battlebanned` variable. We don't have a `.banned` or `.roombanned` variable for the same reason. We can just check whenever the user tries to start a battle.

AustinXII replied:
Kk I’ll fix it although couldn’t you argue the same for ``.locked`` with talking in chats?

panpawn replied:
... locks are a global punishment, I thought we've established that battle bans shouldn't be?

AustinXII replied:
....ok? ``.banned`` is global?? I don’t see your point

AustinXII replied:
I would argue being battlebanned and lock are pretty close to the same severity, this being a global or a room punishment shouldn’t matter imo

panpawn replied:
I meant, global bans and global locks are global punishments, and battle bans aren't

Zarel replied:
`.locked` is more complicated, it's a variable mostly because it comes with a symbol and so filters can set it.

AustinXII replied:
I know @panpawn but it’s irrelevant as ``.banned`` isn't a variable. It seems to only apply to ones that get called a decent amount, which it appears I was mostly correct.

Zarel commented on punishments.js:
```diff
@@ -587,9 +591,9 @@ Punishments.roomPunish = function (room, user, punishment, recursionKeys) {
+		}, roomid + ':' + id, ROOM_PUNISHMENT_FILE);
 
-		if (!(room.isPrivate === true || room.isPersonal || room.battle)) Punishments.monitorRoomPunishments(user);
+		if (roomid === "battle" || !(room.isPrivate === true || room.isPersonal || room.battle)) Punishments.monitorRoomPunishments(user);
```
TypeScript isn't complaining about this? This should probably be `typeof room === 'string' ||`

Zarel commented on punishments.js:
```diff
@@ -1150,6 +1211,25 @@ Punishments.checkName = function (user, userid, registered) {
+			user.cancelReady();
+			if (!punishment) {
+				// Prioritize popups for other global punishments
+				user.send(`|popup|Your are banned from battling${battleban[1] !== userid ? ` because you have the same IP as banned user: ${battleban[1]}` : ``}. Your battle ban will expire in a few days.${battleban[3] ? `||||Reason: ${battleban[3]}` : ``}${Config.appealurl ? `||||Or you can appeal at: ${Config.appealurl}` : ``}`);
```
These popup messages seem to be randomly strewn across the code; they should possibly be consistent.

Also, general admonition about nesting `` ` ``

KrisXV replied:
Also "Your are"

AustinXII replied:
Heh it’s always nice when you arent the one who screwed up, will fix

HoeenCoder commented on punishments.js:
```diff
@@ -593,7 +593,7 @@ Punishments.roomPunish = function (room, user, punishment, recursionKeys) {
 		}, roomid + ':' + id, ROOM_PUNISHMENT_FILE);
 
-		if (roomid === "battle" || !(room.isPrivate === true || room.isPersonal || room.battle)) Punishments.monitorRoomPunishments(user);
+		if (roomid === 'string' || !(room.isPrivate === true || room.isPersonal || room.battle)) Punishments.monitorRoomPunishments(user);
```
Im pretty sure this is supposed to be `typeof roomid === 'string'`

AustinXII replied:
Whoops still need to do the “your” grammar fix too

AustinXII commented:
ugh sorry some of the commit went missing when squashing but yeah should be done

AustinXII commented:
Travis is messing up that message shouldn't have appeared again ^

AustinXII commented:
@Zarel done

Zarel commented:
Also: have you still not figured out how to `npm test`?

AustinXII commented:
? @Zarel  it was fine some of the commit just went missing

Zarel commented:
I'm referring to the 10 Travis failures on this pullreq.

AustinXII commented:
? one was from pushing to the wrong branch and the others were from not knowing the fix to a typescript error which we discussed already :P

panpawn commented on tournaments/index.js:
```diff
@@ -275,7 +275,7 @@ class Tournament {
 		}
 
-		if (this.checkBanned(user)) {
+		if (this.checkBanned(user) || Punishments.isBattleBanned(user)) {
```
What about if someone is already in a tournament and they get battle banned while in the middle of it?

AustinXII replied:
“Prevents the user from starting new battles”  they would be able to finish the match just like how they would be able to if they got tournament banned while in a tournament match

AustinXII replied:
Oh it appears banned users would get kicked, whoops I missed that. Anyways we could just remove the removebanneduser check in checkbanned and move it to this if statement

AustinXII replied:
(Feel free to snipe, I’m just about to start my 10 hour shift)

panpawn replied:
It's a lot more complicated than adding it to this if statement.

You'd have to loop thru all games a user is in (``room.games``), and then check which are tournaments (which I don't actually see a way to do.) Then, if its a tournament, you'd have to create a new function to disqualify the user from it, or I guess you could have them parse ``/tour leave``, but that'd be unreliable if they go offline. 

So yeah, this will be more complicated than a one-liner in an if statement. It should probably be done before the server restarts when this code gets live, though.

AustinXII replied:
Yeah I already figured that out but ty

panpawn replied:
https://github.com/Zarel/Pokemon-Showdown/pull/4468

AustinXII replied:
Completely forgot this still needed fixing on main, ty

Zarel commented on punishments.js:
```diff
@@ -1150,6 +1205,24 @@ Punishments.checkName = function (user, userid, registered) {
 		punishment = Punishments.userids.get(user.locked);
 		if (!punishment) punishment = ['LOCK', user.locked, 0];
 	}
+	if (!battleban && Punishments.isBattleBanned(user)) {
```
Punishments isn't TypeScripted, or else this would have been caught:

This is equivalent to `!battleban && battleban`, which is always `false`.

Zarel replied:
And it's a good thing it's always `false`, because the next two lines are guaranteed to crash.

AustinXII replied:
I think this was copied from the lock command or maybe HH made the change, I don’t really remember writing this part, what was the crash jw

Zarel replied:
`Punishments.isBattleBanned(user)` returns `Punishment | undefined`, which isn't a compatible type to pass to `Punishments.roomUserids.get()`, which expects a `string`.
